### PR TITLE
tsdb: Add gauge histogram support

### PR DESCRIPTION
**Pull request by codesome**
Partially takes care of #11200 (minus scraping the gauge histogram)

In this PR I:
* Added a Type field to Histogram and FloatHistogram, mimicking the headers on the histogram chunks.
* Breaking change to WAL records for this type field.
* Adds support for storing gauge histogram in TSDB. Added relevant methods to see if gauge histogram is appendable, and methods to recode the chunk.

I have broken down the changes into individual PRs and it will be easier to review commit by commit.

### CODE REVIEW

File: model/histogram/generic_test.go [hunk 1]
```diff
@@ -110,3 +110,121 @@ func TestGetBound(t *testing.T) {
 		}
 	}
 }
+
+func TestUnionOfSpans(t *testing.T) {
+	cases := []struct {
+		s1, s2, exp []Span
```
Comment (marctc) [inline model/histogram/generic_test.go:116]: Would it make sense to add a test case where one of the sets is empty?

File: storage/interface.go [hunk 2]
```diff
@@ -282,7 +282,18 @@ type HistogramAppender interface {
...
-	AppendHistogram(ref SeriesRef, l labels.Labels, t int64, h *histogram.Histogram, fh *histogram.FloatHistogram) (SeriesRef, error)
+	//
+	// By default, the storage should assume the histogram to be a counter histogram
+	// unless the hints say it is a gauge histogram.
+	AppendHistogram(ref SeriesRef, l labels.Labels, t int64, h *histogram.Histogram, fh *histogram.FloatHistogram, hints *AppendHints) (SeriesRef, error)
+}
+
+// AppendHints are the hints given to the storage
+// about the sample being appended.
+type AppendHints struct {
```
Comment (marctc) [inline storage/interface.go:293]: Why we pass struct instead just a bool? Do we plan to pass more hints in the future?

File: model/histogram/generic.go [hunk 3]
```diff
@@ -392,6 +392,115 @@ func getBound(idx, schema int32) float64 {
 	return math.Ldexp(frac, exp)
 }
 
+// UnionOfSpans returns a superset of given spans
+// that cover the buckets in both set of spans.
```
Comment (beorn7) [inline model/histogram/generic.go:396]: set → sets

File: storage/interface.go (hunk 2 above)
Comment (beorn7) [reply storage/interface.go:293]: I guess so. We could add hints about counter resets, for example. But @codesome might have had more thoughts.

**Comment by beorn7** [comment]
> * Expanded the `RefFloatHistogramSample` WAL record to store the information of gauge histogram. This is not a breaking change because the float histogram is not in any release yet.

Unfortunately, an integer histogram can be a gauge histogram, too.

Essentially, the whole treatment given to float histograms here has to happen for integer histograms, too.

File: storage/interface.go [hunk 4]
```diff
@@ -282,7 +282,18 @@ type HistogramAppender interface {
 	// For efficiency reasons, the histogram is passed as a
 	// pointer. AppendHistogram won't mutate the histogram, but in turn
 	// depends on the caller to not mutate it either.
-	AppendHistogram(ref SeriesRef, l labels.Labels, t int64, h *histogram.Histogram, fh *histogram.FloatHistogram) (SeriesRef, error)
+	//
+	// By default, the storage should assume the histogram to be a counter histogram
+	// unless the hints say it is a gauge histogram.
+	AppendHistogram(ref SeriesRef, l labels.Labels, t int64, h *histogram.Histogram, fh *histogram.FloatHistogram, hints *AppendHints) (SeriesRef, error)
```
Comment (beorn7) [inline storage/interface.go:288]: Isn't it a bit of an overhead to pass the AppendHints as a pointer? Even if we add more hints to the struct later, it might be leaner to not use a pointer. Which would also avoid ownership problems (if someone recycles the AppendHints).

**Comment by beorn7** [comment]
Very fundamental thought: I have assumed so far that we need a field directly in `model.histogram.Histogram` and `model.histogram.FloatHistogram` that tells us if the histogram is a gauge histogram or not.

Now that we store this information in the TSDB, we should also utilize it in PromQL. I even believe we have to because when storing the result of recording rules, we have to know if the rule created a gauge histogram or a counter histogram. (It will be mostly the former, but nothing prevents a user from just relabeling a counter histogram series and storing the result back to TSDB via a recording rule.)

In fact, my idea was to add an enum field to `model.histogram.Histogram` and `model.histogram.FloatHistogram` that mirrors the already known `CounterResetHeader`, i.e. `CounterReset` / `NotCounterReset` / `GaugeType` / `UnknownCounterReset`. (Float-)Histograms coming from scrapes will either feature `UnknownCounterReset` (for counter histograms) or `GaugeType` (for gauge histograms), but when retrieving (float-)histograms from TSDB, we can set a more specific information about the counter reset. And then it will be almost trivial to utilize the counter reset information in the PromQL engine (which we plan to do anyway, see #11202).

With those ideas, we wouldn't need `AppendHints`, and this PR had to change quite a bit.

Let's perhaps first discuss what path to take before doing a more detailed review.

**Review by beorn7 — COMMENTED** [review]
Just for the record, here a few random tiny remarks that I had already piled up when I realized we have to discuss a few fundamentals first.

File: storage/interface.go (hunk 2 above)
Comment (codesome) [reply storage/interface.go:293]: Yes, more hints in the future possibly. See https://github.com/prometheus/prometheus/issues/11765. It is easier this way where you only have to refactor once.

**Comment by codesome** [comment]
@beorn7 so are you suggesting a breaking change to the current integer histograms?

File: storage/interface.go (hunk 4 above)
Comment (marctc) [reply storage/interface.go:288]: Why adding a pointer adds more overhead?

Comment (beorn7) [reply storage/interface.go]: It means an additional step of de-referencing when accessing the value (only matters in very hot code paths, of course). And it might confuse the escape analysis, maybe triggering the compiler to do a heap allocation where we don't need it.

**Comment by codesome** [comment]
@beorn7 PR is ready for review again. For now iterating the integer histograms always sets UnknownCounterReset on the histogram. And float histogram only tells unknown or gauge. Returning other types during iteration and using it in PromQL can be done in future PRs.

**Comment by beorn7** [comment]
Cool. I'll give it another look ASAP (which is probably tomorrow as my brain feels fried, thinking too much about text format… :exploding_head: )

File: tsdb/chunkenc/histogram_meta.go [hunk 5]
```diff
@@ -230,15 +241,15 @@ type Interjection struct {
 // spans themselves, thanks to the iterators we get to work with the more useful
 // bucket indices (which of course directly correspond to the buckets we have to
 // adjust).
-func compareSpans(a, b []histogram.Span) ([]Interjection, bool) {
+func compareSpans(a, b []histogram.Span, includeBackward bool) (forward, backward []Interjection, ok bool) {
```
Comment (beorn7) [inline tsdb/chunkenc/histogram_meta.go:244]: Since more than half of the code is for the `includeBackward` case, maybe it's easier to create two functions along the lines `forwardCompareSpans` and `bidirectionalCompareSpans` or something…

File: model/histogram/generic.go [hunk 6]
```diff
@@ -392,6 +392,115 @@ func getBound(idx, schema int32) float64 {
 	return math.Ldexp(frac, exp)
 }
 
+// UnionOfSpans returns a superset of given spans
+// that cover the buckets in both sets of spans.
+func UnionOfSpans(s1, s2 []Span) []Span {
```
[the remark this answers is not in the document]
Comment (codesome) [reply model/histogram/generic.go:397]: It is indeed possible to have the union logic inside compareSpans. But I am worried we will create the union for all appends even if we don't need it. Should we consider it later if we see the union being called for large % of appends such that combining both is more beneficial?

File: tsdb/chunkenc/histogram_meta.go (hunk 5 above)
Comment (codesome) [reply tsdb/chunkenc/histogram_meta.go]: Done. With finding merged spans in compareSpans, it makes more sense to split it up.

File: model/histogram/generic.go [hunk 7]
```diff
@@ -392,6 +392,115 @@ func getBound(idx, schema int32) float64 {
...
+		}
+		var currS1, currS2 *Span
+		var s1EndB, s2EndB int32
+
+		if s1idx < len(s1) {
+			currS1 = &s1[s1idx]
+			b1 += currS1.Offset
+			s1idx++
+			if currS1.Length == 0 {
+				// We add consecutive spans until we find a non-zero span.
```
[the remark this answers is not in the document]
Comment (codesome) [reply model/histogram/generic.go]: Not longer needed since this function has been removed

File: model/histogram/float_histogram.go [hunk 8]
```diff
@@ -27,6 +27,8 @@ import (
 // used to represent a histogram with integer counts and thus serves as a more
 // generalized representation.
 type FloatHistogram struct {
+	// CounterResetHint tells the counter reset information for this histogram.
```
Comment (beorn7) [inline model/histogram/float_histogram.go:30]: IIRC the doc comments for fields don't have to start with the name of the field (see other fields in this struct). I'd prefer to have this more compact and less redundant, e.g. "// Counter reset information."

File: model/histogram/histogram.go [hunk 9]
```diff
@@ -35,6 +45,8 @@ import (
 //
 // Which bucket indices are actually used is determined by the spans.
 type Histogram struct {
+	// CounterResetHint tells the counter reset information for this histogram.
```
Comment (beorn7) [inline model/histogram/histogram.go:48]: See above. I'd just say "// Counter reset information."

File: model/histogram/histogram.go [hunk 10]
```diff
@@ -19,6 +19,16 @@ import (
 	"strings"
 )
 
+// CounterResetHint represents the type of a histogram.
```
Comment (beorn7) [inline model/histogram/histogram.go:22]: This still has the "type" nomenclature in the doc comment. Perhaps this is a good location for a slightly more verbose doc comment. E.g. "// CounterResetHint contains the known information about a counter reset, or alternatively that we are dealing with a gauge histogram, where counter resets do not apply."

File: tsdb/chunkenc/float_histogram.go [hunk 11]
```diff
@@ -502,11 +555,29 @@ func (a *FloatHistogramAppender) Recode(
...
+		fh.NegativeBuckets = interject(fh.NegativeBuckets, make([]float64, numNegativeBuckets), nBackwardInter, false)
+	}
+}
+
 type floatHistogramIterator struct {
 	br       bstreamReader
 	numTotal uint16
 	numRead  uint16
 
+	chunkType CounterResetHeader
```
Comment (beorn7) [inline tsdb/chunkenc/float_histogram.go:579]: Maybe rename this, too, along the idea that this is a hint or a piece of information about counter resets rather than a "type". When I read "chunkType", I think this is something about the encoding of the chunk.

File: tsdb/chunkenc/float_histogram.go [hunk 12]
```diff
@@ -211,18 +220,15 @@ func (a *FloatHistogramAppender) AppendHistogram(int64, *histogram.Histogram) {
 // Appendable returns whether the chunk can be appended to, and if so
 // whether any recoding needs to happen using the provided interjections
 // (in case of any new buckets, positive or negative range, respectively).
+// If the sample is a gauge type, AppendableGauge must be used instead.
```
Comment (beorn7) [inline tsdb/chunkenc/float_histogram.go:223]: gauge type → gauge histogram

File: tsdb/chunkenc/float_histogram.go [hunk 13]
```diff
@@ -281,6 +290,50 @@ func (a *FloatHistogramAppender) Appendable(h *histogram.FloatHistogram) (
...
 }
 
+// AppendableGauge returns whether the chunk can be appended to, and if so
+// whether:
+//  1. Any recoding needs to happen to the chunk using the provided interjections
+//     (in case of any new buckets, positive or negative range, respectively).
+//  2. Any recoding needs to happen for the histogram being appended, using the backward interjections
+//     (in case of any missing buckets, positive or negative range, respectively).
+//
+// This method must be only used for gauge type histograms.
```
Comment (beorn7) [inline tsdb/chunkenc/float_histogram.go:300]: gauge type histograms → gauge histograms

File: tsdb/chunkenc/float_histogram.go [hunk 14]
```diff
@@ -211,18 +220,15 @@ func (a *FloatHistogramAppender) AppendHistogram(int64, *histogram.Histogram) {
...
 // • The schema has changed.
-//
 // • The threshold for the zero bucket has changed.
-//
 // • Any buckets have disappeared.
-//
 // • There was a counter reset in the count of observations or in any bucket,
 // including the zero bucket.
-//
 // • The last sample in the chunk was stale while the current sample is not stale.
```
Comment (beorn7) [inline tsdb/chunkenc/float_histogram.go:232]: I think we needed the empty lines because godoc would otherwise munge everything into one paragraph.

By now, godoc supports proper lists, though, see https://pkg.go.dev/github.com/fluhus/godoc-tricks#Lists

So you _can_ remove the empty lines, but you have to replace the `•` with a `-` and indent the lines.

File: tsdb/chunkenc/float_histogram.go [hunk 15]
```diff
@@ -281,6 +290,50 @@ func (a *FloatHistogramAppender) Appendable(h *histogram.FloatHistogram) (
...
+//  2. Any recoding needs to happen for the histogram being appended, using the backward interjections
+//     (in case of any missing buckets, positive or negative range, respectively).
+//
+// This method must be only used for gauge type histograms.
+//
+// The chunk is not appendable in the following cases:
+//
+// • The schema has changed.
+// • The threshold for the zero bucket has changed.
+// • The last sample in the chunk was stale while the current sample is not stale.
```
Comment (beorn7) [inline tsdb/chunkenc/float_histogram.go:306]: See above about lists in godoc.

File: tsdb/head_append.go [hunk 16]
```diff
@@ -1209,24 +1210,37 @@ func (s *memSeries) appendHistogram(t int64, h *histogram.Histogram, appendID ui
...
+		if gauge {
+			positiveInterjections, negativeInterjections, pBackwardInter, nBackwardInter,
+				pMergedSpans, nMergedSpans, okToAppend = app.AppendableGauge(fh)
+		} else {
+			positiveInterjections, negativeInterjections, okToAppend, counterReset = app.Appendable(fh)
+		}
 	}
 
 	if !chunkCreated {
+		if gauge && okToAppend && (len(pBackwardInter) > 0 || len(nBackwardInter) > 0) {
```
Comment (beorn7) [inline tsdb/head_append.go:1239]: `pBackwardInter` and `nBackwardInter` will only ever get filled if both `gauge` and `okToAppend` is `true`.

So you could write this just as `if len(pBackwardInter) + len(nBackwardInter) > 0 {`

**Review by beorn7 — APPROVED** [review]
Looks much better now. I think I even understood it. :o)

A few comment nits and one suggestion about shortening an `if` condition. Otherwise, ship it!

(I have one or two ideas that might improve things even more, but I'm not sure yet. I'll play with them after this is merged, and if the ideas turn out to work, I'll create another PR.)
